test/server/guarddog_impl_test.cc

```diff
@@ -355,6 +438,360 @@ TEST_P(GuardDogTestBase, AtomicIsAtomicTest) {
+
+  void run(envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent event,
+           std::vector<std::pair<Thread::ThreadId, MonotonicTime>> thread_ltt_pairs,
+           MonotonicTime /*now*/) override {
+    std::string event_string =
+        envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent_Name(event);
+    absl::StrAppend(&event_string, " : ");
+    std::vector<std::string> output_string_parts;
+
+    for (const auto& thread_ltt_pair : thread_ltt_pairs) {
+      output_string_parts.push_back(thread_ltt_pair.first.debugString());
+    }
+
+    absl::StrAppend(&event_string, absl::StrJoin(output_string_parts, ","));
+    events_.push_back(event_string);
+  }
+
+protected:
+  std::vector<std::string>& events_; // not owned
+};
+
+// A GuardDogAction that raises the specified signal.
+class SignalGuardDogAction : public Configuration::GuardDogAction {
+public:
+  SignalGuardDogAction(int signal) : signal_(signal) {}
+
+  void run(envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent /*event*/,
+           std::vector<std::pair<Thread::ThreadId, MonotonicTime>> /*thread_ltt_pairs*/,
+           MonotonicTime /*now*/) override {
+    std::raise(signal_);
+  }
+
+protected:
+  const int signal_;
+};
+
+// Test factory for consuming Watchdog configs and creating GuardDogActions.
+template <class ConfigType>
+class TestGuardDogActionFactory : public Configuration::GuardDogActionFactory {
+public:
+  TestGuardDogActionFactory(const std::string& name, std::vector<std::string>& events)
+      : name_(name), events_(events) {}
+
+  Configuration::GuardDogActionPtr createGuardDogActionFromProto(
+      const envoy::config::bootstrap::v3::Watchdog::WatchdogAction& config,
+      Configuration::GuardDogActionFactoryContext& /*context*/) override {
+    // Return different actions depending on the config.
+    if (config.event() == envoy::config::bootstrap::v3::Watchdog::WatchdogAction::KILL) {
+      return std::make_unique<SignalGuardDogAction>(SIGKILL);
+    } else if (config.event() ==
+               envoy::config::bootstrap::v3::Watchdog::WatchdogAction::MULTIKILL) {
+      return std::make_unique<SignalGuardDogAction>(SIGTERM);
+    }
+    return std::make_unique<LogGuardDogAction>(events_);
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return ProtobufTypes::MessagePtr{new ConfigType()};
+  }
+
+  std::string name() const override { return name_; }
+
+  const std::string name_;
+  std::vector<std::string>& events_; // not owned
+};
+
+/**
+ * Tests that various actions registered for the guarddog get called upon.
+ */
+class GuardDogActionsTest : public GuardDogTestBase {
+protected:
+  GuardDogActionsTest() : factory_("TestAction", events_), register_factory_(factory_) {}
+
+  std::vector<std::string> getActionsConfig() {
+    return {
+        R"EOF(
+        {
+          "config": {
+            "name": "TestAction",
```

Comment:
It's odd that all of these test actions have the same name, but act differently.  You may want to omit the event from the set of inputs passed to the factory; the important bits of configuration are the name and the typed config.

Reply:
Well the events are necessary for the `guarddog` to register these actions specified by the config at the appropriate places. I do think the naming is a bit confusing so I'll add another factory that'll generate actions calling `RELEASE_ASSERT(...)` based on one of your previous comments.

Reply:
Yes, the guarddog needs to know the event type for the reasons you describe. The factories/actions don't need to know about the event.